rules/prefer-number-is-integer.js

```diff
@@ -0,0 +1,150 @@
+'use strict';
+
+const MESSAGE_ID = 'preferNumberIsInteger';
+const MESSAGE_ID_SUGGEST = 'preferNumberIsIntegerSuggestion';
+const messages = {
+	[MESSAGE_ID]: 'Replace `{{original}}` with `Number.isInteger({{variable}})`.',
+	[MESSAGE_ID_SUGGEST]: 'Prefer `Number.isInteger()` for integer checks.',
+};
+
+const equalsSelector = ':matches([operator="==="],[operator="=="])';
+
+// Value % 1 === 0
+const modulo1Selector = [
+	'BinaryExpression',
+	'[left.type="BinaryExpression"]',
+	'[left.left.type="Identifier"]',
+	'[left.operator="%"]',
+	'[left.right.value=1]',
+	equalsSelector,
+	'[right.value="0"]',
+].join('');
+
+// (value^0) === value OR (value | 0) === value
+const mathOperatorSelector = [
+	'BinaryExpression',
+	'[left.type="BinaryExpression"]',
+	'[left.left.type="Identifier"]',
+	`:matches(${['^', '|'].map(operator => `[left.operator="${operator}"]`).join(',')})`,
+	'[left.right.value=0]',
+	equalsSelector,
+	'[right.type="Identifier"]',
+].join('');
+
+// ParseInt(value,10) === value
+const parseIntSelector = [
+	'BinaryExpression',
+	'[left.type="CallExpression"]',
+	'[left.callee.name="parseInt"]',
+	'[left.arguments.0.type="Identifier"]',
+	'[left.arguments.1.value=10]',
+	equalsSelector,
+	'[right.type="Identifier"]',
+].join('');
+
+// _.isInteger(value)
+const lodashIsIntegerSelector = [
+	'CallExpression',
+	`:matches(${['_', 'lodash', 'underscore'].map(callee => `[callee.object.name="${callee}"]`).join(',')})`,
+	'[callee.property.name="isInteger"]',
+].join('');
+
+// Math.round(value) === value
+const mathRoundSelector = [
+	'BinaryExpression',
+	'[left.type="CallExpression"]',
+	'[left.callee.type="MemberExpression"]',
+	'[left.callee.object.name="Math"]',
+	'[left.callee.property.name="round"]',
+	'[left.arguments.0.type="Identifier"]',
```

Comment:
I don't think we should limit this to `Identifier`, it should work for anything, just make sure the argument and the right side are the same reference, you can use [`isSameReference`](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/rules/utils/is-same-reference.js).

And don't auto-fix if the expression [has side effect](https://eslint-utils.mysticatea.dev/api/ast-utils.html#hassideeffect).

Reply:
@fisker I've updated the PR to use `isSameReference`.

Can you give me an example of an expression with a side effect that fits here? So I can TDD against it :)